feat: add option to ignore some paths

lucas-bremond opened the pull request:
_Disclaimer: As I am pretty unfamiliar with the Jupytext codebase, this proposal may come across as poorly implemented. Before spending time devising a better implementation and adding tests, I'd like to explore whether the documented intent even makes sense... or not!_

Assuming the following `jupytext.toml` configuration:

```toml
formats = "notebooks///ipynb,source///py:percent"
```

The scenario I've been attempting to tackle is the following.

I have a Jupyter Notebook (`.ipynb`) stored under `notebooks/hello.ipynb`, which thanks to Jupytext gets automatically synced with its Python counterpart at `source/hello.py`.

Now, copying `notebooks/hello.ipynb` over to `export/hello.ipynb` (via the GUI) will fail with the following error:


```
Paste Error

Unexpected error while saving file: export/hello.ipynb Path 'export/hello.ipynb' matches none of the export formats. Please make sure that jupytext.formats covers the current file (e.g. add 'ipynb' to the export formats)
```

which makes sense, since `export/` isn't part of the directory tree "covered" by Jupytext.

In my opinion, an expected behavior in this case could have been to allow the copy resulting in an unpaired notebook.

So the proposed approach here is to introduce an `ignored_paths` option, allowing to define folders that aren't subject to Jupytext's notebook pairing.

Thoughts?

mwouts commented:
Thank you @lucas-bremond for this PR! Yep it makes sense to fix this. I will have a look when time permits.

For now I have the following remarks:
1. It would be great to add a test that reproduces the issue. I can also do that later on (there are many examples in `test_contentsmanager.py`)
2. Is there an error if you create a new notebook in `export` with a `jupytext.toml` file like yours? If so we should address that too (I can have a look when time permits too)
3. Rather than ignoring the `export` directory, I would rather make sure that there is no pairing / no error if the notebook is moved/created in another directory than `notebooks` or `source`.

lucas-bremond commented:
Hello @mwouts, many apologies for the belated reply.

I have rebased my PR on top of the latest `main` branch, and added tests as requested (however, I was only able to test the `rename_file` flow, and not the `copy` flow; if you have a good pattern for this please let me know!).

> Is there an error if you create a new notebook in export with a jupytext.toml file like yours? If so we should address that too (I can have a look when time permits too)

I've just tested, and there is no error there.

> Rather than ignoring the export directory, I would rather make sure that there is no pairing / no error if the notebook is moved/created in another directory than notebooks or source.

Indeed... what would be the recommended approach (this likely would require a deeper understanding of the codebase)?

lucas-bremond commented:
👋 Hello @mwouts! Anything I could do/improve to make this mergeable? :)

mwouts commented:
Hi @lucas-bremond , thank you for the ping, I am sorry I must have missed your previous message. I will finalize the review and merge this week-end hopefully. Quick question about the precedence of the ignore config vs the metadata: I would have thought that the notebook metadata should take precedence over the global config, would that be fine with you?

mwouts commented:
Hi @lucas-bremond , I was testing your PR (reverting the changes on the metadata as discussed above at https://github.com/mwouts/jupytext/compare/pr_1199).

I am wondering it if would not be preferable to address the problem you reported in the first place - moving a notebook outside of its current paired folder should be possible. Also, this might date back to a while ago but I guess that you removed the jupytext metadata for a reason, do you remember what it could be?

lucas-bremond commented:
Hi @mwouts, apologies for the belated reply.

> I am wondering it if would not be preferable to address the problem you reported in the first place - moving a notebook outside of its current paired folder should be possible.

Certainly. I'm assuming removing the additional `ignored_paths` setting, and dropping the pairing information by default? I didn't do it that way to not potentially impact any existing flow (hence the "opt-in" option), but your call!

> Also, this might date back to a while ago but I guess that you removed the jupytext metadata for a reason, do you remember what it could be?

I believe this was because I wanted notebooks copied to the destinations listed in `ignored_paths` to come out as "regular notebooks" (clean of Jupytext pairing instructions). I was afraid that otherwise, Jupytext would try to re-create the paired notebook.

mwouts commented:
Hi @lucas-bremond , I have prepared a PR that should allow moving notebooks outside of what the formats says, could you give it a try and let me know if it solves the issue ? Instructions for testing are at https://github.com/mwouts/jupytext/pull/1422#issuecomment-3194660539. Thanks!

mwouts commented:
I have merged #1422 which allows moving a notebook, and will now release v1.17.3, please let me know how it works for you!

lucas-bremond commented:
Apologies for the delay @mwouts, I've tested it and it works perfectly! Thank you so much.

mwouts commented:
> Apologies for the delay @mwouts, I've tested it and it works perfectly! Thank you so much.

Excellent, thank you for confirming! No worries at all
